### PR TITLE
LightningCLI documentation improvements

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

It includes several improvements to the documentation of `LightningCLI`. There is now a section that explains the use of command line arguments as discussed in #8093. It also explains the use of defaults for parameters that have a class as type hint. And includes some notes about reproducibility.

The version of jsonargparse is also bumped. With this now it is possible to configure subclasses via individual command line arguments without having to create a dictionary.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by mauvilsa** [comment]
@jlperla

File: docs/source/common/lightning_cli.rst [hunk 1]
```diff
@@ -196,6 +194,58 @@ Note that there is a section for each class (model and trainer) including all th
 grouping is also used in the formatting of the help shown previously.
 
 
+Use of command line arguments
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For every CLI implemented, users are encouraged to learn how run it by reading the documentation printed with the
```
Comment (tchaton) [inline docs/source/common/lightning_cli.rst:200]:
```suggestion
For every CLI implemented, users are encouraged to learn how to run it by reading the documentation printed with the
```

File: docs/source/common/lightning_cli.rst [hunk 2]
```diff
@@ -196,6 +194,58 @@ Note that there is a section for each class (model and trainer) including all th
...
+:code:`--help` option and use the :code:`--print_config` option to guide the writing of config files. A few more details
+that might not be clear by only reading the help are the following.
+
+:class:`~pytorch_lightning.utilities.cli.LightningCLI` is based on argparse and as such follows the same arguments style
+as many POSIX command line tools. Long options are prefixed with two dashes and its corresponding values can be given
+separated by a space or an equal sign, as :code:`--option value` or :code:`--option=value`. The same options can be
```
Comment (tchaton) [inline docs/source/common/lightning_cli.rst:206]:
```suggestion
:class:`~pytorch_lightning.utilities.cli.LightningCLI` is based on argparse and as such follows the same arguments style
as many POSIX command line tools. Long options are prefixed with two dashes and its corresponding values should be provided with an empty space or an equal sign, as :code:`--option value` or :code:`--option=value`. The same options can be
```

File: docs/source/common/lightning_cli.rst [hunk 3]
```diff
@@ -196,6 +194,58 @@ Note that there is a section for each class (model and trainer) including all th
...
+given multiple times and ones given later (to the right) override the previous ones. If a class has an init parameter
+that is required, it also has to be given as :code:`--option` which makes things explicit and more readable instead of
+relying on positional arguments.
+
+When calling a CLI all options can be provided using individual arguments. However, given the large amount of options
```
Comment (tchaton) [inline docs/source/common/lightning_cli.rst:211]:
```suggestion
When calling a CLI, all options can be provided using individual arguments. However, given the large amount of options
```

File: docs/source/common/lightning_cli.rst [hunk 4]
```diff
@@ -196,6 +194,58 @@ Note that there is a section for each class (model and trainer) including all th
...
+.. code-block:: bash
+
+    $ python trainer.py --config experiment_defaults.yaml --trainer.max_epochs 100
+
+Another common pattern could be having multiple config files, two examples being:
+
+.. code-block:: bash
+
+    $ python trainer.py --config config1.yaml --config config2.yaml [...]
+    $ python trainer.py --trainer trainer.yaml --model model.yaml --data data.yaml [...]
```
Comment (tchaton) [inline docs/source/common/lightning_cli.rst:225]: Neat !

File: docs/source/common/lightning_cli.rst [hunk 5]
```diff
@@ -535,16 +646,12 @@ A corresponding example of the config file would be:
...
-      ...
-    trainer:
-      ...
 
 And the same through command line:
 
 .. code-block:: bash
 
-    $ python train.py --optimizer='{class_path: torch.optim.Adam, init_args: {lr: 0.01}}'
+    $ python train.py --optimizer.class_path=torch.optim.Adam --optimizer.init_args.lr=0.01
```
Comment (carmocca) [inline docs/source/common/lightning_cli.rst:654]: Nice!

**Review by carmocca — COMMENTED** [review]
Thanks a lot!

File: docs/source/common/lightning_cli.rst (hunk 5 above)
Comment (jlperla) [reply docs/source/common/lightning_cli.rst:654]: Great!

**Comment by jlperla** [comment]
Given the above `python train.py --optimizer.class_path=torch.optim.Adam --optimizer.init_args.lr=0.01` I assume that if I define other defaults in a separate file that these will override them.  If so, documenting it might be very valuable.  Would it be something liike
```bash
python trainer.py --config default_config.yaml -optimizer.class_path=torch.optim.Adam --optimizer.init_args.lr=0.01
```
If so, then it would be great to show people that (with the example a config that would have the optimizer stuff setup that they could edit).  

Then the thing I really would love to see in in the docs is something telling me how to make that default config file a default, so I don't need to pass it in each time!  Then I can just register `./default_config.yaml` as the default in my subclassed Lightning Parser somewhere, make sure it is in that folder, then just go
```bash
python trainer.py --optimizer.class_path=torch.optim.Adam --optimizer.init_args.lr=0.01
```
or to use all of the defaults!
```bash
python trainer.py
```
where I can then still use any other config file with `python trainer.py --config another_config.yml`

File: docs/source/common/lightning_cli.rst [hunk 6]
```diff
@@ -196,6 +194,58 @@ Note that there is a section for each class (model and trainer) including all th
...
+
+For every CLI implemented, users are encouraged to learn how run it by reading the documentation printed with the
+:code:`--help` option and use the :code:`--print_config` option to guide the writing of config files. A few more details
+that might not be clear by only reading the help are the following.
+
+:class:`~pytorch_lightning.utilities.cli.LightningCLI` is based on argparse and as such follows the same arguments style
+as many POSIX command line tools. Long options are prefixed with two dashes and its corresponding values can be given
+separated by a space or an equal sign, as :code:`--option value` or :code:`--option=value`. The same options can be
+given multiple times and ones given later (to the right) override the previous ones. If a class has an init parameter
+that is required, it also has to be given as :code:`--option` which makes things explicit and more readable instead of
```
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source/common/lightning_cli.rst:207]: This relates to a parameter without a default, for example in:

```python
def my_function(
    param1: str,
    param2: str = 'default'
):
    ...
```

`param1` is required. Currently in jsonargparse there are only two possibilities when adding a signature like this as arguments: `as_positional=False` or `as_positional=True`. If configured to be positional then `param1` can only be given as positional, not with its named option. In fire this kind of arguments can be given as positional or optional and surely would be a nice feature to have in jsonargparse. Not sure how difficult it is to implement. Anyway currently this is how it works.

**Comment by mauvilsa** [comment]
> Given the above `python train.py --optimizer.class_path=torch.optim.Adam --optimizer.init_args.lr=0.01` I assume that if I define other defaults in a separate file that these will override them. If so, documenting it might be very valuable. Would it be something liike
> 
> ```shell
> python trainer.py --config default_config.yaml -optimizer.class_path=torch.optim.Adam --optimizer.init_args.lr=0.01
> ```

This is already explained. There is a whole section for it with examples.

> Then the thing I really would love to see in in the docs is something telling me how to make that default config file a default, so I don't need to pass it in each time!

Will add this.

**Comment by mauvilsa** [comment]
> > Then the thing I really would love to see in in the docs is something telling me how to make that default config file a default, so I don't need to pass it in each time!
> 
> Will add this.

Done

**Comment by kaushikb11** [comment]
@mauvilsa I believe there has been a bad merge.

**Comment by mauvilsa** [comment]
> @mauvilsa I believe there has been a bad merge.

I have no idea what I did. But now it is fixed.

**Review by carmocca — APPROVED** [review]

File: docs/source/common/lightning_cli.rst [hunk 7]
```diff
@@ -196,6 +194,74 @@ Note that there is a section for each class (model and trainer) including all th
...
+
+In some cases people might what to pass an entire config in an environment variable, which could also be used instead of
+a path to a file, for example:
+
+.. code-block:: bash
+
+    $ python trainer.py --trainer "$TRAINER_CONFIG" --model "$MODEL_CONFIG" [...]
+
+An alternative for environment variables could be to instantiate the CLI with :code:`env_parse=True`. In this case the
+the help shows the names of the environment variables for all options. A global config would be given in
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:245]:
```suggestion
An alternative for environment variables could be to instantiate the CLI with :code:`env_parse=True`. In this case the help shows the names of the environment variables for all options. A global config would be given in
```
word repetition

File: docs/source/common/lightning_cli.rst [hunk 8]
```diff
@@ -196,6 +194,74 @@ Note that there is a section for each class (model and trainer) including all th
...
+:code:`PL_CONFIG` and there wouldn't be a need to specify any command line argument.
+
+It is also possible to set a path to a config file of defaults. If the file exists it would be automatically loaded
+without having to specify any command line argument. Arguments given would override the values in the default config
+file. To load a defaults file :code:`my_cli_defaults.yaml` in current working directory would be implemented as:
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:250]:
```suggestion
It is also possible to set a path to a config file of defaults. If the file exists it would be automatically loaded
without having to specify any command line argument. Arguments given would override the values in the default config
file. Loading a defaults file :code:`my_cli_defaults.yaml` in the current working directory would be implemented as:
```

File: docs/source/common/lightning_cli.rst [hunk 9]
```diff
@@ -416,6 +482,67 @@ To change the configuration of the :code:`EarlyStopping` in the config it would
...
+            self,
+            backbone: torch.nn.Module = MyModel(encoder_layers=24),  # BAD PRACTICE!
+        ):
+            super().__init__()
+            self.backbone = backbone
+
+Normally classes are mutable as it is in this case. The instance of :code:`MyModel` would be created the moment that the
+module that defines :code:`MyMainModel` is first imported. This means that the default of :code:`backbone` will be
+initialized before the CLI class runs :code:`seed_everything` making it non-reproducible. Furthermore, if
+:code:`MyMainModel` is used more than once in the same python process and the :code:`backbone` parameter is not
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:514]:
```suggestion
:code:`MyMainModel` is used more than once in the same Python process and the :code:`backbone` parameter is not
```

File: docs/source/common/lightning_cli.rst [hunk 10]
```diff
@@ -585,3 +708,23 @@ a single class or a tuple of classes, the value given to :code:`optimizer_init`
...
+people can use in unexpected ways. Nevertheless :class:`~pytorch_lightning.utilities.cli.LightningCLI` tries to give a
+framework and recommendations to make reproducibility simpler.
+
+When an experiment is run, it is good practice to use a stable version of the source code, either being a released
+package or at least a commit of some version controlled repository. For each run of a CLI the config file is
+automatically saved including all settings. This is useful to figure out what was done for a particular run without
+requiring to look at the source code. If by mistake the exact version of the source code is lost or some defaults
+changed, having the full config means that most of the information is preserved.
+
+The class is targeted at implementing CLIs because running a command from a shell provides a separation with the python
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:726]:
```suggestion
The class is targeted at implementing CLIs because running a command from a shell provides a separation with the Python
```

**Review by awaelchli — APPROVED** [review]

File: docs/source/common/lightning_cli.rst [hunk 11]
```diff
@@ -93,13 +93,11 @@ practice to create a configuration file and provide this to the tool. A way to d
     nano config.yaml
     # Run training using created configuration
     python trainer.py --config config.yaml
-    # The config JSON can also be passed directly
-    python trainer.py --config '{trainer: {fast_dev_run: True}}'
```
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source/common/lightning_cli.rst:97]: Not really discourage. Actually this is still documented where the environment variables are explained.

File: docs/source/common/lightning_cli.rst [hunk 12]
```diff
@@ -196,6 +194,74 @@ Note that there is a section for each class (model and trainer) including all th
...
+
+.. code-block:: bash
+
+    $ python trainer.py --config experiment_defaults.yaml --trainer.max_epochs 100
+
+Another common pattern could be having multiple config files, two examples being:
+
+.. code-block:: bash
+
+    $ python trainer.py --config config1.yaml --config config2.yaml [...]
```
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source/common/lightning_cli.rst:224]: If you understood this then I need to rewrite a bit to make it clearer. Both config1 and config2 are picked. All arguments are always parsed in order from left to right. If inside both config1 and config2 there is a common setting, then the one used is the one in config2. But also settings from config1 will be used if not in config2.

[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source/common/lightning_cli.rst:224]: Rewrote it. Please check.

[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source/common/lightning_cli.rst:224]: thanks

**Review by kaushikb11 — APPROVED** [review]
